### PR TITLE
AndroidHIDNetProvider and Android Sample

### PR DESCRIPTION
**Pull request by MelbourneDeveloper**
This pull request adds an Android sample.  I did a slight refactor where I moved a parameter from the constructor of HIDNetDevice up to WindowsHIDNetDevice. I hope this is OK. Android is just a bit of a different beast. But, if this is no good, I can probably change this.

Thanks for the refactoring! It wasn't clear how it worked at first, but I think it's a great step forward. It decouples the library from the underlying Hid layer. I am working towards removing the need to DataHasExtraByte, and I will work on implementing a full AndroidHIDNetProvider. As you said, the plan is to keep track of a list of devices that have been attached/detached. Good idea.

Please working with you.

### CODE REVIEW

File: LedgerWalletAndroidSample/AndroidHIDNetDevice.cs [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+        }
+
+        public override IHIDDevice Clone()
+        {
+            throw new NotImplementedException();
+        }
+
+        public override Task EnsureInitializedAsync(CancellationToken cancellation)
+        {
+            throw new NotImplementedException();
```
Comment (NicolasDorier) [inline LedgerWalletAndroidSample/AndroidHIDNetDevice.cs]: if not needed, remove it.

File: LedgerWalletAndroidSample/AndroidHIDNetDevice.cs [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+        AndroidHidDevice _AndroidHidDevice;
+
+        public AndroidHIDNetDevice(AndroidHidDevice androidHidDevice) : base(androidHidDevice)
+        {
+            _AndroidHidDevice = androidHidDevice;
+        }
+
+        public override IHIDDevice Clone()
+        {
+            throw new NotImplementedException();
```
Comment (NicolasDorier) [inline LedgerWalletAndroidSample/AndroidHIDNetDevice.cs]: just make a close, `new AndroidHIDNetDevice(_AndroidHidDevice)`

**Comment by NicolasDorier** [comment]
You have not used the `IHidProvider` facility at all, so one can't just use `LedgerClient.GetHIDDevices()`, this make the library different to use on different plateform, not good.

The idea is that the consumer of the lib just have to do something like this:

```
HidProvider.Provider = new AndroidHIDProvider();
```
Then later...

```
var devices = await LedgerClient.GetHIDDevicesAsync();
```

The fact that you can't enumerate the devices with the API is not a problem. You can make the AndroidHIDProvider listen the plug/unplug events directly and maintain the list of device at this level.

**Comment by MelbourneDeveloper** [comment]
This was the first step. I will implement the provider fully but didn't have enough time.

**Comment by MelbourneDeveloper** [comment]
@NicolasDorier I have added the AndroidHIDNetProvider as requested.

**Comment by NicolasDorier** [comment]
This is not how it is meant to be used.

The provider need to register to android events and save the hid device informations in a list which it returns. That way can simplify the life for the users of the lib by not asking them to do it.

Basically, all the code in `MainActivity.cs` related to listening android events should be gone and moved to AndroidHIDProvider.

**Comment by MelbourneDeveloper** [comment]
@NicolasDorier , unfortunately I've run out of time to work on this. It's going to be some time before I can change this and resubmit it.

**Comment by MelbourneDeveloper** [comment]
@NicolasDorier , unfortunately I've run out of time to work on this. It's going to be some time before I can change this and resubmit it.

**Comment by NicolasDorier** [comment]
thanks, I might take a look later.

**Comment by MelbourneDeveloper** [comment]
@NicolasDorier 

> The provider need to register to android events and save the hid device informations in a list which it returns. That way can simplify the life for the users of the lib by not asking them to do it.

I understand why this is what you want to achieve. You want to put a layer across all code so that the user does not have to write bifurcated code across the platforms. This makes total sense.

I made some attempts at achieving this but mostly it just broke my Android sample. I have come to think that Hid.Net also needs a layer on top which puts a similar provider type system over all platforms. This will be very difficult because connecting to devices on each platform is very idiosyncratic and debugging is very difficult - especially without unit testing. It generally takes me around five minutes to run and test a code change for Android.

I can't commit to any short term goals, but if you want to wait, I will eventually put a layer in Hid.Net that will simplify connecting to devices in a uniform way (or at least as uniform as possible). Then, the provider implementation for the Ledger .NET API will be a lot simpler and more reliable. However, I can't promise any timeframes on this right now.

**Pull request closed without merging**